src/main/java/sparksoniq/spark/iterator/flowr/WhereClauseSparkIterator.java

```diff
@@ -20,30 +20,103 @@
+import sparksoniq.jsoniq.item.Item;
 import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
 import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
 import sparksoniq.jsoniq.runtime.tupleiterator.RuntimeTupleIterator;
+import sparksoniq.jsoniq.runtime.tupleiterator.SparkRuntimeTupleIterator;
 import sparksoniq.jsoniq.tuple.FlworTuple;
+import sparksoniq.semantics.DynamicContext;
 import sparksoniq.spark.closures.WhereClauseClosure;
 import sparksoniq.spark.iterator.flowr.base.FlowrClauseSparkIterator;
 
-public class WhereClauseSparkIterator extends FlowrClauseSparkIterator {
+import javax.naming.OperationNotSupportedException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class WhereClauseSparkIterator extends SparkRuntimeTupleIterator {
+
+    private RuntimeIterator _expression;
+    private DynamicContext _tupleContext;   // re-use same DynamicContext object for efficiency
+    private FlworTuple _nextLocalTupleResult;
+    private FlworTuple _inputTuple;     // tuple received from child, used for tuple creation
+
     public WhereClauseSparkIterator(RuntimeTupleIterator child, RuntimeIterator whereExpression, IteratorMetadata iteratorMetadata) {
-        super(child, null, FLWOR_CLAUSES.WHERE, iteratorMetadata);
-        this._children.add(whereExpression);
+        super(child, iteratorMetadata);
+        _expression = whereExpression;
+    }
+
+    @Override
+    public boolean isRDD() {
+        return _child.isRDD();
+    }
+
+    @Override
+    public void open(DynamicContext context) {
+        super.open(context);
+
+        // isRDD checks omitted, as open is used for non-RDD(local) operations
+
+        if (this._child != null) {
+            _child.open(_currentDynamicContext);
+            _tupleContext = new DynamicContext(_currentDynamicContext);     // assign current context as parent
+
+            setNextLocalTupleResult();
+
+        } else {
+            throw new SparksoniqRuntimeException("Invalid where clause.");
+        }
+    }
+
+    @Override
+    public FlworTuple next() {
+        if(_hasNext == true){
+            FlworTuple result = _nextLocalTupleResult;      // save the result to be returned
+            setNextLocalTupleResult();              // calculate and store the next result
+            return result;
+        }
+        throw new IteratorFlowException("Invalid next() call in let flwor clause", getMetadata());
+    }
+
+    private void setNextLocalTupleResult() {
+        // for each incoming tuple, evaluate the expression to a boolean.
+        // forward if true, drop if false
+
+        while (_child.hasNext()) {
+            _inputTuple = _child.next();
+            _tupleContext.removeAllVariables();             // clear the previous variables
+            _tupleContext.setBindingsFromTuple(_inputTuple);      // assign new variables from new tuple
+
+            _expression.open(_tupleContext);
+            Item result = _expression.next();
```

Comment:
This is slightly more complicated than that -- you need to compute the Effective Boolean Value of the sequence, as done at other places.

In particular, you need to test for _expression.hasNext() that may be false (empty sequence, in which the EBV is false).

Then you need to test, if _expression.next() is an atomic item, that it is a singleton sequence: if _expression.hasNext() then returns true after next() was an atomic, then an error must be raised. If next() is an object or array, no need to call hasNext() a second time as the EBV is then always true.

It is very likely that this is already implemented at other places and that you do not need to reinvent any new code, for example you can check in the implementation of the IfThenElse iterator, for the computation of the condition.

Reply:
Actually I checked the implementation of IfThenElse and it only checks for one item. You may need to refactor and move over getEffectiveBoolean Value from the item class to the iterator class, and have it take an item iterator as a parameter (rather than an item), and returns a boolean or raises an error.